packages/dev/core/src/AudioV2/abstractAudio/subNodes/abstractAudioSubGraph.ts

```diff
@@ -0,0 +1,146 @@
+import type { Nullable } from "../../../types";
+import type { AbstractAudioNode, AbstractNamedAudioNode } from "../abstractAudioNode";
+import type { _AbstractAudioSubNode } from "./abstractAudioSubNode";
+
+/**
+ * Adds common sub graph functionality to an audio node.
+ *
+ * Audio nodes such as static sounds, streaming sounds, and buses can use audio sub graphs to process audio internally
+ * before sending it to connected downstream audio nodes. This is useful for applying effects, spatial audio, and other
+ * audio processing tasks common to multiple audio node classes.
+ *
+ * A key feature of audio sub graphs is their audio sub nodes are created asynchronously on demand so the minimum set
+ * of sub nodes are used at all times to save memory and CPU resources. The tradeoff is a small delay when first
+ * setting a property backed by a sub node. This delay is avoided by using the appropriate options to initialize the
+ * sub node on creation, e.g. `spatialEnabled` and `stereoEnabled`, or by setting any creation option backed by the
+ * sub node, e.g. `spatialPosition` and `stereoPan`.
+ *
+ * @internal
+ */
+export abstract class _AbstractAudioSubGraph {
+    private _createSubNodePromises: { [key: string]: Promise<_AbstractAudioSubNode> } = {};
+    private _subNodes: { [key: string]: Set<AbstractNamedAudioNode> } = {};
```

Comment:
Should probably use Map for these.

Reply:
Raanan suggested using this to save memory.